File: sdks/typescript/src/gateway.ts
import type { DecisionRecordPayload } from "./types.js";
import type { VelarixSession } from "./client.js";

export class VelarixGateway {
  private session: VelarixSession;
  private mode: "strict" | "buffered";
  private buffer: DecisionRecordPayload[] = [];
  private maxBuffered: number;

  constructor(session: VelarixSession, opts?: { mode?: "strict" | "buffered"; maxBuffered?: number }) {
    this.session = session;
    this.mode = opts?.mode ?? "strict";
    this.maxBuffered = typeof opts?.maxBuffered === "number" && opts.maxBuffered > 0 ? Math.floor(opts.maxBuffered) : 2000;
  }

  async flush(): Promise<void> {
    if (this.buffer.length === 0) return;
    const pending = [...this.buffer];
    this.buffer = [];
    for (const p of pending) {
      await this.session.recordDecision(p);
    }
  }

  private async record(payload: DecisionRecordPayload): Promise<void> {
    try {
      if (this.mode === "buffered") await this.flush();
      await this.session.recordDecision(payload);
    } catch (e) {
      if (this.mode !== "buffered") throw e;
      if (this.buffer.length >= this.maxBuffered) {
        // Fail closed once buffering limit is hit.
        throw e;
      }
      this.buffer.push(payload);
    }
  }

  async callTool<TInput extends Record<string, any>, TOutput>(
    tool: string,
    input: TInput,
    fn: (input: TInput) => Promise<TOutput>,
    opts?: { traceId?: string; tags?: string[] },
  ): Promise<TOutput> {
    const trace_id = opts?.traceId;
    const tags = opts?.tags;

    await this.record({
      kind: "tool_call",
      trace_id,
      tool,
      input,
      tags,
    } as DecisionRecordPayload);

    try {
      const output = await fn(input);
      await this.record({
        kind: "tool_result",
        trace_id,
        tool,
        input,
        output,
        tags,
      } as DecisionRecordPayload);
      return output;
    } catch (e: any) {
      await this.record({
        kind: "error",
        trace_id,
        tool,
        input,
        error: String(e?.message || e),
        tags,
      } as DecisionRecordPayload);
      throw e;
    }
  }
}

File: sdks/python/velarix/gateway.py
from typing import Any, Callable, Dict, Optional


class VelarixGateway:
    """Record tool activity to Velarix while tolerating transient write failures."""

    def __init__(self, session, mode: str = "strict", max_buffered: int = 2000):
        self.session = session
        self.mode = mode or "strict"
        self.max_buffered = max(1, int(max_buffered))
        self._buffer = []

    def flush(self) -> None:
        """Replay any buffered decision events through the attached session."""
        if not self._buffer:
            return
        pending = list(self._buffer)
        self._buffer = []
        for kind, payload in pending:
            self.session.record_decision(kind, payload)

    def _record(self, kind: str, payload: Dict[str, Any]) -> None:
        try:
            if self.mode == "buffered":
                self.flush()
            self.session.record_decision(kind, payload)
        except Exception:
            if self.mode != "buffered":
                raise
            if len(self._buffer) >= self.max_buffered:
                raise
            self._buffer.append((kind, payload))

    def call_tool(
        self,
        tool: str,
        input: Dict[str, Any],
        fn: Callable[[Dict[str, Any]], Any],
        trace_id: Optional[str] = None,
        tags: Optional[list] = None,
    ) -> Any:
        """Record a tool call, execute it, and persist either the result or the error."""
        self._record("tool_call", {"trace_id": trace_id, "tool": tool, "input": input, "tags": tags or []})
        try:
            output = fn(input)
            self._record(
                "tool_result",
                {"trace_id": trace_id, "tool": tool, "input": input, "output": output, "tags": tags or []},
            )
            return output
        except Exception as e:
            self._record("error", {"trace_id": trace_id, "tool": tool, "input": input, "error": str(e), "tags": tags or []})
            raise
